src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -378,69 +365,131 @@ and gen_from_arrow ~loc ~env left right =
         let res, xs = aux xs in
         let obs = observable x in
         (res, [%expr [%e obs] @-> [%e xs]])
-    | x -> (gen_from_type ~loc ~env x, [%expr o_nil])
+    | x -> (gen_from_type ~loc  x, [%expr o_nil])
   in
   let x, obs = aux right in
+  (* TODO: export this in qcheck_generators for https://github.com/c-cube/qcheck/issues/190 *)
   let arb = [%expr QCheck.make [%e x]] in
   [%expr
     QCheck.fun_nary QCheck.Tuple.([%e observable left] @-> [%e obs]) [%e arb]
     |> QCheck.gen]
 
-let gen_from_kind_variant ~loc ~env typ_name xs =
+let gen_from_kind_variant ~loc xs =
   let (module A) = Ast_builder.make loc in
-  let is_rec (constr : constructor_declaration) : bool =
-    match constr.pcd_args with
-    | Pcstr_tuple xs -> List.exists (is_rec_typ typ_name) xs
-    | _ -> false
-  in
-  sized ~loc ~env typ_name is_rec (gen_from_constr ~loc) xs
+  let is_rec cd = is_rec_constr_decl (curr_types ()) cd in
+  sized ~loc is_rec (gen_from_constr ~loc) xs
+
+(** [gen_from_type_declaration loc td] creates a generator from the type declaration.
 
-let rec curry_args ~loc args body =
-  match args with
-  | [] -> body
-  | x :: xs -> [%expr fun [%p x] -> [%e curry_args ~loc xs body]]
+    It returns either `Recursive or `Normal.
 
-let gen_from_type_declaration ~loc ?(env = TypeGen.empty) td =
-  let name = td.ptype_name.txt in
-  let pat_gen = pat ~loc name in
+    - `Normal of expression:
+    The derivated generator is not recursive, we return only the generator.
```

Comment:
Minor: The *derived* generator

Reply:
Fixed in 696b9b9aecbcdeaa75b1324f2ab75f0f813630f2.